### PR TITLE
Allow excluding files when discovering

### LINKED ISSUE #59: `discover` pragma: Allow exclusion of modules
The `expand` pragma of cabal-fmt allows the exclusion of certain modules: `-- cabal-fmt: expand directory/ -ModuleToExclude`.
It was nice if cabal-gild could do that too, since the `discover` pragma cannot be used if there are modules in the directory tree that should have mixed visibilities (`other-modules` vs `exposed-modules`). For example, if we have a directory tree similar to the following:
```
src/Foo/Bar.hs
src/Foo/Baz.hs
src/Foo/Helper.hs
```
Here, if `Foo.Helper` is supposed to be in `other-modules` while the other ones are supposed to be in `exposed-modules`, then we cannot use module discovery for `src`.

Also, the lack of this feature prevents module discovery to be used in test suites or executable components:
```
test-suite test
  type: exitcode-stdio-1.0
  hs-source-dirs: test
  main-is: Main.hs

  -- cabal-fmt: expand test/ -Main
  -- cabal-gild: discover test/
  other-modules:
```
Here, cabal-gild will add the `Main` module to the `other-modules` field. But this package description cannot be built by cabal:
```
    module ‘main:Main’ is defined in multiple files: test/Main.hs
                                                     test/Main.hs
```

### PR DESCRIPTION
**Pull request by tfausak**
Fixes #59. 

This makes two changes to the `discover` pragma:

- Previously at least one directory was required. Now no directories are required. If no directories are given, it will discover modules from the current directory. This is the same as saying `-- cabal-gild: discover .`.
- There is a new flag: `--exclude=MODULE`. This can be given to exclude modules from the discovery process. For example you might say `-- cabal-gild: discover --exclude Foo.Internal` to avoid discovering the `Foo.Internal` module. 

Note that (for now) the `--exclude=MODULE` option only accepts a single module name. You can provide it multiple times to exclude multiple modules. However it does not support any type of wildcards.

### CODE REVIEW

**Comment by tfausak** [comment]
This pull request does not add any support for globbing. However, in the future Gild might add the ability to exclude modules based on a glob. For example it might be convenient to say `--exclude=**.Internal` or `--exclude=**.*Spec` or `--exclude=Company.**`.

**Comment by tfausak** [comment]
Looking for feedback from @mmhat, @konn & @nikita-volkov. Would these changes work for you? I think globs are an obvious next step, but module exclusions should be useful by themselves.

**Comment by tfausak** [comment]
Another natural extension of this is adding support for `--include`. I think that only makes sense if both it and `--exclude` support globs. Otherwise module inclusion isn't any better than just explicitly listing the modules (and avoiding discovery). 

Supporting both inclusion and exclusion would let you do stuff like this:

``` cabal
library
  -- cabal-gild: discover --exclude **.Internal
  exposed-modules: ...
  -- cabal-gild: discover --include **.Internal
  other-modules: ...
```

**Comment by konn** [comment]
Looks good, thanks! I think this at least solves the problem with tasty-discover, as we can now write as:

```cabal
test-suite tests
  type: exitcode-stdio-1.0
  hs-source-dirs: test
  main-is: Spec.hs
  -- cabal-gild: discover test --exclude=Spec.hs
  other-modules:
```

**Comment by tfausak** [comment]
Note that as currently implemented, the `--exclude` option takes a list of _modules_, not a list of _files_. That means you would want to do `--exclude=Spec`.

This is one of the things I'm unsure about though. Should it accept files instead of modules? I think most of the time people are going to be matching against `something.hs`, so removing that `.hs` suffix is nice. Also it's nice to not have to deal with path separators across platforms (since modules use `.` as separators). But I think Cabal uses `/` as a path separator for "symbolic" paths across platforms, so maybe that doesn't matter. 

However matching on files is strictly more powerful. The difference doesn't matter much right now, but if globbing were supported then maybe you'd want to do something like this:

``` cabal
-- cabal-gild: discover --exclude=**/*.hsig
exposed-modules: ...

-- cabal-gild: discover --exclude=**/*.hs
signatures: ...
-- or --include=**/*.hsig
```

That is, automatically discover both exposed modules and signatures in the same directory, distinguishing them by their extensions. 

Matching on files also opens the door to share this functionality with other fields like `data-files` or `extra-source-files`. Those already natively support globs, but perhaps people want to resolve the globs when formatting the package description rather than when building the package. (For example, see https://github.com/haskell/cabal/issues/8075.)

**Comment by mmhat** [comment]
Hi @tfausak , thanks for picking this up; I ran out of time when I was working on a fix.

Yes, this solution would work for me, although I'd prefer one using file paths for the reasons you outlined in your previous comments.

Apart from `Spec.hs` modules I use this functionality to exclude the `main-is:` module. I believe we could even exclude that by default as Cabal complains if it is listed in `other-modules:`, and if we were doing that then exclusion by file path is easier.

**Comment by tfausak** [comment]
Thanks for the feedback! It seems to me that the best way forward is for `--exclude` to take a file path rather than a module name. That fixes the immediate problem by allowing `--exclude=Spec.hs`. And it leaves the door open for accepting globs (probably via [filepattern](https://hackage.haskell.org/package/filepattern-0.1.3)). And then also adding `--include`. And finally supporting other fields like `data-files` that end up with files instead of modules.

File: README.md [hunk 1]
```diff
@@ -207,3 +208,14 @@ before a field.
...
   `*.x`, or `*.y`,
+
+  Directories can be quoted if they contain spaces.
+
+  Discovered modules can be ignored by using the `--exclude=MODULE` option. For
+  example:
+
+  ``` cabal
+  library
+    -- cabal-gild: discover source/library --exclude=Foo.Bar
```
Comment (tfausak) [inline README.md]:
```suggestion
  Discovered modules can be ignored by using the `--exclude=FILE` option. For
  example:

  ``` cabal
  library
    -- cabal-gild: discover source/library --exclude=source/library/Foo/Bar.hs
```

File: source/test-suite/Main.hs [hunk 2]
```diff
@@ -1,12 +1,16 @@
+{- hlint ignore "Redundant bracket" -}
 {-# LANGUAGE GeneralizedNewtypeDeriving #-}
```
Comment (tfausak) [inline source/test-suite/Main.hs:2]:
```suggestion
{-# LANGUAGE GeneralizedNewtypeDeriving #-}
```

**Comment by tfausak** [comment]
Ok, I think I'm happy with this for now. The only potentially confusing thing is that the excluded files are given relative to the package description. So for example you'd need to do this:

``` hs
-- cabal-gild: discover test --exclude=test/Spec.hs
```

Rather than just `--exclude=Spec.hs`. That's because `discover` can take any number of directories. Just `Spec.hs` would be ambiguous in a case like this:

``` hs
-- cabal-gild: discover foo bar --exclude=Spec.hs
```

**Pull request merged**